Mathlib/RingTheory/Polynomial/Hilbert.lean

```diff
@@ -0,0 +1,199 @@
+/-
+Copyright (c) 2024 Fangming Li. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fangming Li, Jujian Zhang
+-/
+import Mathlib.Algebra.Polynomial.RingDivision
+import Mathlib.RingTheory.PowerSeries.WellKnown
+
+/-!
+# Hilbert polynomials
+
+Given any `p : ℤ[X]` and `d : ℕ`, there exists some `h : ℚ[X]` such that for any
+large enough `n : ℕ`, `PowerSeries.coeff ℤ n (p * (@invOneSubPow ℤ _ d))` is equal
+to `h.eval (n : ℚ)`. This `h` is unique and is called the Hilbert polynomial with
+respect to `p` and `d` (`Polynomial.hilbert p d`).
+-/
+
+open BigOperators
+open PowerSeries
+
+namespace Polynomial
+
+section greatestFactorOneSubNotDvd
+
+variable {R : Type*} [CommRing R] [NeZero (1 : R)] [NoZeroDivisors R]
+variable (p : R[X]) (hp : p ≠ 0) (d : ℕ)
+
+/--
+Given a polynomial `p`, the greatest factor of `p` that is not divided by `1 - X`.
```

Comment:
```suggestion
Given a polynomial `p`, the greatest factor of `p` that is not divisible by `1 - X`.
```
This docstring should be more precise, explaining which sign you are choosing here. Factors are only defined up to units.